contracts/BC_fusion/BSCGovernor.sol

```diff
@@ -0,0 +1,381 @@
+import "@openzeppelin/contracts-upgradeable/governance/GovernorUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/governance/compatibility/GovernorCompatibilityBravoUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/governance/extensions/GovernorVotesUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/governance/extensions/GovernorTimelockControlUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/governance/extensions/GovernorVotesQuorumFractionUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/governance/extensions/GovernorPreventLateQuorumUpgradeable.sol";
+
+import "./System.sol";
+import "./lib/Utils.sol";
+import "./interface/IGovToken.sol";
+
+contract BSCGovernor is
+    System,
+    Initializable,
+    GovernorUpgradeable,
+    GovernorSettingsUpgradeable,
+    GovernorCompatibilityBravoUpgradeable,
+    GovernorVotesUpgradeable,
+    GovernorTimelockControlUpgradeable,
+    GovernorVotesQuorumFractionUpgradeable,
+    GovernorPreventLateQuorumUpgradeable
+{
+    using Utils for bytes;
+    using Utils for string;
+
+    /*----------------- constants -----------------*/
+    /**
+     * @dev caution:
+     * INIT_VOTING_DELAY, INIT_VOTING_PERIOD and INIT_MIN_PERIOD_AFTER_QUORUM are default in number of blocks, not seconds
+     */
+    uint256 private constant BLOCK_INTERVAL = 3 seconds;
+    uint256 private constant INIT_VOTING_DELAY = 0 hours / BLOCK_INTERVAL;
+    uint256 private constant INIT_VOTING_PERIOD = 7 days / BLOCK_INTERVAL;
+    uint256 private constant INIT_PROPOSAL_THRESHOLD = 200 ether; //  = 200 BNB
+    uint256 private constant INIT_QUORUM_NUMERATOR = 10; // for >= 10%
+
+    // starting propose requires totalSupply of GovBNB >= 10000000 * 1e18
+    uint256 private constant PROPOSE_START_GOVBNB_SUPPLY_THRESHOLD = 10_000_000 ether;
+    // ensures there is a minimum voting period (1 days) after quorum is reached
+    uint64 private constant INIT_MIN_PERIOD_AFTER_QUORUM = uint64(1 days / BLOCK_INTERVAL);
+
+    /*----------------- errors -----------------*/
+    // @notice signature: 0x584a7938
+    error NotWhitelisted();
+    // @notice signature: 0x11b6707f
+    error TotalSupplyNotEnough();
+    // @notice signature: 0xe96776bf
+    error GovernorPaused();
+    // @notice signature: 0x286300de
+    error OnlyGovernorProtector();
+    // @notice signature: 0xb1d02c3d
+    error InBlackList();
+    // @notice signature: 0x867f3ee5
+    error OneLiveProposalPerProposer();
+
+    /*----------------- events -----------------*/
+    event Paused();
+    event Resumed();
+    event BlackListed(address indexed target);
+    event UnBlackListed(address indexed target);
+
+    /*----------------- storage -----------------*/
+    // target contract => is whitelisted for governance
+    mapping(address => bool) public whitelistTargets;
+
+    bool public proposeStarted;
+    bool public paused;
```

Comment:
Is there any `Paused` interfact/library, several contracts have very similar requirement on pause - they can share codes.

Reply:
Done